File: scala/median-of-two-sorted-arrays/src/main/scala/com/leetcode/scala/Area.scala
package com.leetcode.scala

class Area[T](val arr: Array[T], val start: Double = Double.NaN,
  val end: Double = Double.NaN) {

  private var _section = new Section(head = _start, tail = _end)
  private lazy val _arr = arr
  private lazy val _start = start
  private lazy val _end = end

  def section = _section

  def head: T = {
    apply(_start.toInt)
  }

  def head(section: Section): T = {
    apply(section.start.toInt)
  }

  def tail: T = {
    apply(_end.toInt)
  }

  def tail(section: Section): T = {
    apply(section.end.toInt)
  }

  def apply(index: Int): T = {
    if (!(index >= _start && index <= _end))
      throw IllegalArgumentException(
        "Index should within the interval [start, end]")
    else
      arr.apply(index)
  }

  def :=(section: Section): this.type = {
    _section = section
    this
  }

  def :=(start: Double = -1, end: Double = -1): this.type = {
    _section = new Section(head = start, tail = end)
    this
  }
}

File: javascript/median-of-two-sorted-arrays/src/area.js
import IllegalArgumentException from './illegal-argument-exception';
import NullSetException from './null-set-exception';
import Section from './section';

/**
 * 表示区间数据类
 * @class
 */
class Area {
  constructor(options = { arr: [], section: null }) {
    const { getInstances } = this.constructor;
    const { arr, section } = options;

    // 按顺序可配置和不可配置属性设置默认值
    Object.assign(this, options, getInstances());

    // 缓存数组长度
    this.arrLen = this.arr.length;

    // 如果只有`arr`变量,就根据其初始化`section`
    if (section == null) {
      this.initSection();
    } else {
      this.section = section;
    }
  }

  /**
   * @private
   * @static
   * @method
   * @return {Object} 用户不能配置的实例属性默认值
   */
  static getInstances() {
    return {
      arrLen: 0,
    };
  }

  /**
   * @public
   * @method
   */
  initSection() {
    const { arr: { length } } = this;

    this.section = new Section({ start: 0, end: length - 1 });
  }

  /**
   * 返回区域首个元素
   * @public
   * @method
   */
  getHead() {
    const { arr, section } = this;
    const { start } = section;
    let head

    section.checkNull();
    if (start >= 0) {
      head = arr[start];
    } else {
      throw new InvalidArgumentException(
        'Area first element index should lesser and equals than zero;');
    }

    return head;
  }

  /**
   * 返回区域末尾元素
   * @public
   * @method
   */
  getTail() {
    const { arr, arrLen, section } = this;
    const { end } = section;
    let tail;

    section.checkNull();
    if (end < arrLen) {
      tail = arr[end];
    } else {
      throw new InvalidArgumentException(
        'Area last element index should lesser than arr length;');
    }

    return tail;
  }

  /**
   * 遍历区间
   * @public
   * @method
   */
  forEach(callback) {
    const { arr, section } = this;
    const { start, end } = section;

    try {
      section.checkNull();
    } catch (e) {
      return false;
    }

    for (let index = start; index <= end; index++) {
      callback(arr[index], index, arr);
    }
  }


  /**
   * 获取区间总长度
   * @public
   * @method
   */
  getTotal() {
    return this.section.length;
  }

  /**
   * 获取区间总长度
   * @public
   * @method
   */
  getClip(section =  null) {
    const { arr } = this;
    const area = new Area({ arr, section });

    return area;
  }
}

export default Area;
